### PR TITLE
pythonPackages.cachy: 0.2.0 -> 0.3.0

### PR DESCRIPTION
**Pull request by jonringer**
###### Motivation for this change
closes #67043

Added tests to make sure I wasn't breaking anything when removing version bound. Unfortunately, poetry is self-bootstrapping (i.e. need a previous version to install it), I had to do both a pypi fetch for the installation, and a github fetch for the tests.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

```
[4 built, 7 copied (1.7 MiB), 0.4 MiB DL]
https://github.com/NixOS/nixpkgs/pull/72445
6 package were build:
poetry python27Packages.cachy python27Packages.poetry python37Packages.cachy python38Packages.cachy python38Packages.poetry
```

### CODE REVIEW

**Comment by jonringer** [comment]
@GrahamcOfBorg build poetry python27Packages.cachy python27Packages.poetry python37Packages.cachy python38Packages.cachy python38Packages.poetry

File: pkgs/development/python-modules/poetry/default.nix [hunk 1]
```diff
@@ -29,11 +33,20 @@ let
...
-  glob2 = callPackage ./glob2.nix { };
 
+  # used for tests, can't be used directly for install because
+  # it's missing a setup.py
+  githubSrc = fetchFromGitHub {
+    owner = "sdispater";
+    repo = "poetry";
+    rev = version;
+    sha256 = "004s747wkil5f00r0vjff73r6vhlrrcnfan2k7pl7gyf62wfp02a";
+  };
```
Comment (risicle) [inline pkgs/development/python-modules/poetry/default.nix]: If needs must, I guess - at least both package fetches are based on the same `version` so they shouldn't diverge.

**Comment by risicle** [comment]
`nox-review` happy for me macos 10.13 & non-nixos linux x86_64

File: pkgs/development/python-modules/poetry/default.nix (hunk 1 above)
Comment (jonringer) [reply pkgs/development/python-modules/poetry/default.nix]: yes, which is why i moved the version to the the let scope...

**Comment by jonringer** [comment]
@GrahamcOfBorg build poetry python27Packages.cachy python27Packages.poetry python37Packages.cachy python38Packages.cachy python38Packages.poetry

not sure why darwin's tests are still all passing, but then the drv still fails to build....

File: pkgs/development/python-modules/poetry/default.nix (hunk 1 above)
Comment (jonringer) [reply pkgs/development/python-modules/poetry/default.nix]: only way I thought i could sneak tests in... didn't want to break it

Comment (jonringer) [reply pkgs/development/python-modules/poetry/default.nix]: plus, there's apparently no packages that use poetry to build... so i can't use them as a way to verify correctness :(

**Comment by jonringer** [comment]
oh python38Packages.rednose is broken, the other's are good

@FRidh is this too hacky for you? :)

File: pkgs/development/python-modules/poetry/default.nix (hunk 1 above)
Comment (FRidh) [reply pkgs/development/python-modules/poetry/default.nix]: the source can be used by using `format = "pyproject";`

Comment (jonringer) [reply pkgs/development/python-modules/poetry/default.nix]: the raw source of poetry can't be used without a previous version of poetry, the github pull is just to get the tests.

**Review by FRidh — APPROVED** [review]

**Review by FRidh — CHANGES_REQUESTED** [review]
actually wait I think it can be done...

**Comment by FRidh** [comment]
@jonringer how hacky is this? :-D
https://github.com/NixOS/nixpkgs/commit/48749ff32f45f1955c4f0d2beb9bfa3a1c982e1e

**Comment by jonringer** [comment]
I'll try again by modifying the python path, might be possible

**Comment by jonringer** [comment]
I feel like that would make it a pain to update :)

**Comment by FRidh** [comment]
I don't know. As long as poetry works directly from the source directory, I don't see any problem with it. Replacing versions is essentially the same as using `substituteInPlace` or `sed`. The big advantage here is that we're building straight from VCS. Recently I did the same for `setuptools`, `pip` and `wheel`, because that way you can test e.g. dev versions. Now, I don't think it's likely anyone will do that for `poetry`, at least not as long as its a relatively unknown project.

**Comment by FRidh** [comment]
Regarding poetry again. In the future pip will support self-hosting backends, so the `PYTHONPATH` part can be removed as soon as `poetry` implements it.
https://discuss.python.org/t/pep-517-backend-bootstrapping/789

**Pull request closed without merging**